src/catchup/ApplyCheckpointWork.cpp

```diff
@@ -265,15 +268,37 @@ ApplyCheckpointWork::onRun()
             openInputFiles();
         }
 
-        auto result = applyHistoryOfSingleLedger();
         auto const& lm = mApp.getLedgerManager();
+        auto& bl = mApp.getBucketManager().getBucketList();
+
+        bl.resolveAnyReadyFutures();
+        if (!bl.futuresAllResolved(
+                bl.getMaxMergeLevel(lm.getLastClosedLedgerNum() + 1)))
+        {
+            std::weak_ptr<ApplyCheckpointWork> weak(
+                std::static_pointer_cast<ApplyCheckpointWork>(
+                    shared_from_this()));
+            auto handler = [weak](asio::error_code const& ec) {
+                auto self = weak.lock();
+                if (self)
+                {
+                    self->wakeUp();
+                }
+            };
+
+            // Check back later
+            mDelayTimer.expires_from_now(std::chrono::milliseconds(500));
```

Comment:
it would be cleaner to use `ConditionalWork` for this and introduce a `applyLedgerWork`.
By doing this you will share the code to apply a ledger here and in the "buffered ledger" case

Reply:
When doing this, add a better status message for `ConditionalWork` (maybe something like `Waiting (X for Y seconds) before Z` where:
X is the name of the condition
Y is the time elapsed since the conditional work was started
Z is the status of the conditioned work

Reply:
It's not clear to me what code will be shared here. The "buffered ledger" case just calls `closeLedger` to apply the ledger, while `ApplyCheckpointWork` calls `applyHistoryOfSingleLedger`, which has logic specific to `ApplyCheckpointWork`. Are you saying that logic should be shared?

Reply:
yeah what I am saying is imagine that you have a simple "wait and apply ledger" work (that takes as a parameter `LedgerCloseData`):
* then you would use this directly in the buffered ledger case
* in `applyHistoryOfSingleLedger` you would only invoke it as a "sub-work" only after it passes certain pre-conditions and you would check some post-conditions to potentially cause `ApplyCheckpointWork` to fail